ykrt/pass_finder/try_passes.py

```diff
@@ -1,6 +1,11 @@
 #!/usr/bin/env python3
-
-import argparse, math, os, shutil, sys, time, queue 
+"""
+This script is designed to identify the sequences of passes that successfully
+build the yklua interpreter and execute its tests. If a sequence fails, the
+algorithm divides it into two sections and retries. The order of the passes, as
+initially received, is preserved throughout the process.
+"""
```

Comment:
We can tweak this to be more precise:

 - It's not specific to `ylklua`: you can use any oracle.
 - We need to say what happens after the sequence is chopped in two?
 - We need to say what the starting sequence is.

Reply:
Does following sounds better?

> This script aims to identify the sequences of passes that successfully builds and executes oracle's tests 10 times. The algorithm starts with a list of passes such as [pass1, pass2, pass3, pass4, pass5], if a sequence fails to build and/or successfully run the oracle's test, the algorithm divides the list into two subsets (i.e [pass1, pass2] and [pass3, pass4, pass5])  and retries to build and run oracle's tests 10 times . The algorithm parallely process the subsets and has a limit of 10 process to prevent fork bomb. The order of the passes, as initially received, is preserved throughout the process.

Reply:
Much better. I'll propose some tweaks/restructuring:

> This script aims to identify an LLVM AOT pipeline that satisfies a test oracle. A typical test oracle would be a script that builds an interpreter and runs a test suite. This script is required because some passes don't yet work with Yk.
>
> The algorithm starts with the default clang `-O2` pipeline which is sent to the test oracle for evaluation. If the oracle accepts the pipeline (it returns zero), then the passes are appended to an "accept list".  If the oracle rejects the pipeline (it returns non-zero), then the algorithm divides the failing pipeline into two parts and tests the two halves independently (in parallel). The search continues recursively until the search space is exhausted. Upon termination the final accept list is printed.

Now the only things we don't talk about are:
 - The inputs to the oracle. IIRC it's an environment variable?
 - The pre-link and post-link pipelines, and their role in all of this.

Can you try an edit to include those details?

[I expect we will go back and forth a few more times on this]

Reply:
Updated, should I talk more about what changes are made in the pipeline (i.e changes in `cflags` and `ldflags`)?

> usage: python3 try_passes.py [-lto] [-prelink] -- path path/to/test/oracle
> This script aims to identify an LLVM AOT pipeline that satisfies a test oracle. A typical test oracle would be a script that builds an interpreter and runs a test suite. This script is required because some passes don't yet work with Yk.
> The algorithm expects the default pre and/or post-link clang `-O2` pipeline which is sent to the test oracle for evaluation. After receiving a list of prelink and postlink passes, the test oracle sets an environment variable to enable these passes, overriding the default pipeline. If the oracle accepts the pipeline (it returns zero), then the passes are appended to an "accept list".  If the oracle rejects the pipeline (it returns non-zero), then the algorithm divides the failing pipeline into two parts and tests the two halves independently (in parallel). The search continues recursively until the search space is exhausted. Upon termination the final accept list is printed.

Reply:
> The algorithm expects the default pre and/or post-link clang -O2 pipeline

I don't know what you mean by "expects" here. Don't you query `opt` to get the starting -O2 pipeline?

> pre and/or post-link clang -O2 pipeline

A user will ask: when is it "and" and when is it"or"?

> the test oracle sets an environment variable to enable these passes, overriding the default pipeline

The script doesn't actually enable or disable any passes itself, it just sets an environment variable and it's up to the oracle to ensure that those passes are respected. I think that need to be clearer.

> should I talk more about what changes are made in the pipeline (i.e changes in cflags and ldflags)?

We might not need that here, because it's the job of the oracle.

Sorry if this is annoying, but you can see the level of precision we are striving for.

Reply:
> > The algorithm expects the default pre and/or post-link clang -O2 pipeline
> 
> I don't know what you mean by "expects" here. Don't you query `opt` to get the starting -O2 pipeline?

> 
> > pre and/or post-link clang -O2 pipeline

> A user will ask: when is it "and" and when is it"or"?

I meant the algorithm "expects" user to specify  ***at least*** one option (i.e `lto` or `prelink`) and then query `opt` to start the pipeline. If it doesn't receieve either one of the two options then the script will end with an error.

> > the test oracle sets an environment variable to enable these passes, overriding the default pipeline
> 
> The script doesn't actually enable or disable any passes itself, it just sets an environment variable and it's up to the oracle to ensure that those passes are respected. I think that need to be clearer.

Yes, so rephrasing the line to:
> After receiving a list of prelink and postlink passes, the test oracle sets an environment variable, overriding the default pipeline


> Sorry if this is annoying, but you can see the level of precision we are striving for.

no problem.

Reply:
> I meant the algorithm "expects" user to specify

OK, so we can say: "the user specifies..."

Which reminds me, we should add a usage line. You know when the tool prints stuff like:

```
usage: the_tool [-v] this_arg that_arg
```

> "After receiving a list of prelink and postlink passes, the test oracle sets an environment variable, overriding the default pipeline"

Good. Even better(?):

> Each time the oracle is invoked it is passed a list of prelink and postlink passes in the environment variable `BLAH`. The oracle must ensure that these pipelines used when compiling the interpreter (usually by passing the pipeline to the `--pre-link-pipeline` and `--post-link-pipeline` arguments of `yk-config`.

(something like that? adjust as you see fit)

Reply:
Great. This is very close now.

> usage: python3 try_passes.py [-lto] [-prelink] -- path path/to/test/oracle

Are the `-lto` and `-prelink` arguments "one or the other, but one is required"? If so I'd format it like `<-lto|-prelink>`.

> path/to/test/oracle

I thought the path to the oracle was hard-coded for now?

Also, please wrap the docstring at 80 characters.

Reply:
No, the path to interepreter is expected, the test file name is hardcodes. So we are using `test.sh` but the directory path is given by user.

Reply:
OK, ideally later we should make it so the user specifies the whole path to the script.